lib/migrate.js

```diff
@@ -0,0 +1,444 @@
+// This file is licensed under the Apache License 2.0.
+// License text available at https://opensource.org/licenses/Apache-2.0
+
+'use strict';
+
+var async = require('async');
+var assert = require('assert');
+var g = require('strong-globalize')();
+var util = require('util');
+var _ = require('lodash');
+
+module.exports = mixinMigrate;
+
+function mixinMigrate(CouchDB) {
+  var debug = require('debug')('loopback:connector:couchdb:migrate');
+
+/**
+ * Perform automigrate for the given models.
+ * - destroys model data if the model exists
+ * - delete existing indexes in database
+ * - create new indexes according to model definition
+ * 
+ * @param {String|String[]} [models] A model name or an array of model names.
+ * If not present, apply to all models
+ * @callback {Function} cb The callback function
+ */
+CouchDB.prototype.automigrate = function(models, cb) {
+  debug('CouchDB.prototype.automigrate models %j', models);
+  var self = this;
+  var existingModels = models;
+  async.series([
+    function(callback) {
+      destroyData(callback);
+    },
+    function(callback) {
+      debug('CouchDB.prototype.automigrate update indexes for %j', models);
+      self.migrateOrUpdateIndex(models, true, callback);
+    }], cb);
+  function destroyData(destroyCb) {
+    async.eachSeries(existingModels, function(model, cb) {
+      self.destroyAll(model, {}, {}, cb);
+    }, function(err) {
+      debug('CouchDB.prototype.automigrate destroy all data: %j', err);
+      destroyCb(err);
+    });
+  };
+};
+
+/**
+ * Perform autoupdate for the given models.
+ * It does NOT destroy previous model instances if model exists, only
+ * `automigrate` does that.
+ *  - compare new indexes and existing indexes, 
+ *  - keep unchanged indexes
+ *  - add newly defined indexes
+ *  - delete old indexes  
+ * @param {String[]} [models] A model name or an array of model names. If not
+ * present, apply to all models
+ * @callback {Function} cb The callback function
+ */
+
+CouchDB.prototype.autoupdate = function(models, cb) {
+  this.migrateOrUpdateIndex(models, false, cb);
+}
+
+/**
+ * Add and delete certain indexes, performs depends on which function calls it.
+ * 
+ * @param {String[]} [models] A model name or an array of model names. Passed in
+ * from `automigrate` or `autoupdate`
+ * @param {boolean} isMigrate 
+ * - `true` when called by `automigrate`
+ * - `false` when called by `autoupdate`
+ * 
+ * @callback {Function} cb The callback function
+ */
+
+CouchDB.prototype.migrateOrUpdateIndex = function(models, isMigrate, cb) {
+  debug('CouchDB.prototype.migrateOrUpdateIndex %j', models);
+
+  var self = this;
+  async.each(models, autoUpdateOneModel, cb);
+
+  function autoUpdateOneModel(model, cb) {
+    debug('CouchDB.prototype.migrateOrUpdateIndex updating model: %j', model);
+    var mo = self.selectModel(model, true);
+    assert(mo, 'model ' + model + ' does not exist in your registry!');
+    var indexes = mo.mo.model.definition.indexes();
```

Comment:
@b-admike If you look at the code in https://github.com/strongloop/loopback-connector-couchdb2/blob/master/lib/couchdb.js#L716-L723

the `mo` in this line is actually the whole object get returned from code above, which contains `mo` as its property, true the naming looks confusing